File: system-test/dotnet/Dsl.Core/UseCase/MyShopDsl.cs
using Driver.Port;
using Dsl.Core.UseCase.UseCases;
using Dsl.Core.Shared;

namespace Dsl.Core.UseCase;

public class MyShopDsl : IAsyncDisposable
{
    private readonly IMyShopDriver _driver;
    private readonly UseCaseContext _context;

    private MyShopDsl(IMyShopDriver driver, UseCaseContext context)
    {
        _driver = driver;
        _context = context;
    }

    public static Task<MyShopDsl> CreateAsync(IMyShopDriver driver, UseCaseContext context)
    {
        return Task.FromResult(new MyShopDsl(driver, context));
    }

    public async ValueTask DisposeAsync()
    {
        if (_driver != null)
            await _driver.DisposeAsync();
    }

    public GoToMyShop GoToMyShop() => new(_driver, _context);

    public PlaceOrder PlaceOrder() => new(_driver, _context);

    public CancelOrder CancelOrder() => new(_driver, _context);

    public DeliverOrder DeliverOrder() => new(_driver, _context);

    public ViewOrder ViewOrder() => new(_driver, _context);

    public PublishCoupon PublishCoupon() => new(_driver, _context);

    public BrowseCoupons BrowseCoupons() => new(_driver, _context);
}

File: system/multitier/backend-java/src/testSupport/java/com/mycompany/myshop/backend/support/core/usecase/MyShopDsl.java
package com.mycompany.myshop.backend.support.core.usecase;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.mycompany.myshop.backend.support.BackendDriver;
import com.mycompany.myshop.backend.support.core.usecase.usecases.BrowseCoupons;
import com.mycompany.myshop.backend.support.core.usecase.usecases.BrowseOrderHistory;
import com.mycompany.myshop.backend.support.core.usecase.usecases.GoToMyShop;
import com.mycompany.myshop.backend.support.core.usecase.usecases.PlaceOrder;
import com.mycompany.myshop.backend.support.core.usecase.usecases.PublishCoupon;
import com.mycompany.myshop.backend.support.core.usecase.usecases.ViewOrder;

/**
 * The system under test, one class per use case. Every call produces an outcome the caller states an
 * expectation on ({@code .execute().shouldSucceed()} / {@code .shouldFail()}) — unlike the external
 * stub DSLs, whose {@code execute()} is fire-and-forget because programming an in-process WireMock
 * cannot fail.
 */
public class MyShopDsl {

    private final BackendDriver driver;
    private final ObjectMapper objectMapper;

    public MyShopDsl(BackendDriver driver, ObjectMapper objectMapper) {
        this.driver = driver;
        this.objectMapper = objectMapper;
    }

    public GoToMyShop goToMyShop() {
        return new GoToMyShop(driver);
    }

    public PlaceOrder placeOrder() {
        return new PlaceOrder(driver, objectMapper);
    }

    public ViewOrder viewOrder() {
        return new ViewOrder(driver, objectMapper);
    }

    public BrowseOrderHistory browseOrderHistory() {
        return new BrowseOrderHistory(driver);
    }

    public PublishCoupon publishCoupon() {
        return new PublishCoupon(driver);
    }

    public BrowseCoupons browseCoupons() {
        return new BrowseCoupons(driver);
    }
}
